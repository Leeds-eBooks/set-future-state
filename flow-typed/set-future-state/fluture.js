// @flow

type Cancel = () => void
type Nodeback<E, V> = (error: ?E, value?: V) => void

declare class Future<E, V> {
  map<M>(f: (value: V) => M): Future<E, M>;
  mapRej<M>(f: (error: E) => M): Future<M, V>;

  fork(onReject: (error: E) => *, onResolve: (value: V) => void): Cancel;
  done(callback: Nodeback<E, V>): Cancel;
}

declare module 'fluture' {
  declare export function isFuture(f: Future<*, *>): true
  declare export function isFuture(f: mixed): false

  declare export function tryP<V>(p: () => Promise<V>): Future<*, V>
  declare export function node<E, V>(
    n: (callback: Nodeback<E, V>) => void
  ): Future<E, V>

  declare export function encaseN<A, E, V>(
    n: (a: A, callback: Nodeback<E, V>) => void,
    a: A
  ): Future<E, V>
  declare export function encaseN<A, E, V>(
    n: (a: A, callback: Nodeback<E, V>) => void
  ): (a: A) => Future<E, V>

  declare export function of<V>(value: V): Future<null, V>
  declare export function after<V>(ms: number, value: V): Future<null, V>
  declare export function rejectAfter<E>(ms: number, error: E): Future<E, void>

  declare export function Future<E, V>(
    (reject: (error: E) => void, resolve: (value: V) => void) => ?Cancel
  ): Future<E, V>
}
