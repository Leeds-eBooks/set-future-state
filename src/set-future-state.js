// @flow

import {Component, PureComponent} from 'react'
import {isFuture, tryP} from 'fluture'

const getErrorTarget = ({constructor}) => {
  const name = constructor.displayName || constructor.name
  return name ? ` Check your ${name} component.` : ''
}

const create = SuperClass =>
  class FutureState<P, S> extends SuperClass<P, S> {
    _cancels: Array<() => void> = []

    // FIXME: prevent (or handle) overriding â€“ monkey-patch? Inject?
    componentWillUnmount() {
      this._cancels.forEach(fn => fn())
    }

    setFutureState<V>(
      eventual: Future<V> | (() => Promise<V>),
      reducer: (value?: V, prevState: S, props: P) => $Shape<S> | null,
      onError?: Error => mixed
    ) {
      if (!(isFuture(eventual) || typeof eventual === 'function')) {
        throw new TypeError(
          `The first argument to setFutureState() must be a Future or a Function which returns a Promise.${getErrorTarget(
            this
          )}`
        )
      }

      const future = typeof eventual === 'function' ? tryP(eventual) : eventual

      const resolver = (v?: V) =>
        this.setState((prevState, props) => reducer(v, prevState, props))

      this._cancels.push(
        onError
          ? future.fork(onError, resolver)
          : future.done((e, v) => resolver(v))
      )
    }
  }

export class ComponentFutureState<P, S> extends create(Component)<P, S> {}
export class PureComponentFutureState<P, S> extends create(PureComponent)<
  P,
  S
> {}
