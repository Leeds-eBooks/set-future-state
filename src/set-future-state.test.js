// @flow

import React from 'react'
import {of, after, rejectAfter, Future as future} from 'fluture'
import renderer from 'react-test-renderer'
import {PureComponentFutureState} from './set-future-state'

const wait = (ms: number): Promise<void> =>
  new Promise(resolve => setTimeout(resolve, ms))

type Props = {
  // flowlint-next-line unclear-type:off
  eventual: Future<any> | (() => Promise<any>),
  reducer: *,
  onError?: *,
}

type State = {
  loading: boolean,
  v?: mixed,
  props?: {},
}

class Test extends PureComponentFutureState<Props, State> {
  state = {loading: false}

  trigger() {
    this.setFutureState(
      this.props.eventual,
      this.props.reducer,
      this.props.onError
    )
  }

  render() {
    return this.state.loading.toString()
  }
}

describe('PureComponentFutureState', () => {
  describe('setFutureState', () => {
    it('throws for an incorrect first argument', () => {
      // $FlowExpectError
      const test = <Test eventual={''} reducer={() => ({})} />
      const render = renderer.create(test)
      const instance = render.getInstance()
      expect(instance.trigger.bind(instance)).toThrowErrorMatchingSnapshot()

      render.update(
        React.cloneElement(test, {eventual: () => Promise.resolve()})
      )
      expect(instance.trigger.bind(instance)).not.toThrow()

      render.update(React.cloneElement(test, {eventual: of('')}))
      expect(instance.trigger.bind(instance)).not.toThrow()
    })

    describe('onError', () => {
      const onError = jest.fn()
      const test = (
        <Test
          eventual={() => Promise.reject(new Error('TEST'))}
          reducer={() => ({})}
          onError={onError}
        />
      )
      let render
      let instance

      beforeEach(() => {
        onError.mockClear()
        render = renderer.create(test)
        instance = render.getInstance()
      })

      it('is called for a rejected Promise', async () => {
        expect(onError).not.toHaveBeenCalled()
        instance.trigger()
        expect(onError).not.toHaveBeenCalled()
        await wait(3)
        expect(onError).toHaveBeenCalledWith(expect.any(Error))
      })

      it('is called for a rejected Future', async () => {
        render.update(
          React.cloneElement(test, {
            eventual: rejectAfter(1, new Error('TEST')),
          })
        )

        expect(onError).not.toHaveBeenCalled()
        instance.trigger()
        expect(onError).not.toHaveBeenCalled()
        await wait(3)
        expect(onError).toHaveBeenCalledWith(expect.any(Error))
      })
    })

    describe('reducer', () => {
      const reducer = jest.fn()
      const test = (
        <Test eventual={() => Promise.resolve(5)} reducer={reducer} />
      )
      let render
      let instance

      beforeEach(() => {
        reducer.mockClear()
        render = renderer.create(test)
        instance = render.getInstance()
      })

      describe('with Promises', () => {
        it('is called for a resolved with undefined onError', async () => {
          expect(reducer).not.toHaveBeenCalled()
          instance.trigger()
          expect(reducer).not.toHaveBeenCalled()
          await wait(2)
          expect(reducer).toHaveBeenCalledWith(
            5,
            {loading: false},
            {eventual: expect.any(Function), reducer}
          )
        })

        it('is called for a resolved with defined onError', async () => {
          render.update(React.cloneElement(test, {onError: () => null}))

          expect(reducer).not.toHaveBeenCalled()
          instance.trigger()
          expect(reducer).not.toHaveBeenCalled()
          await wait(2)
          expect(reducer).toHaveBeenCalledWith(
            5,
            {loading: false},
            {
              eventual: expect.any(Function),
              reducer,
              onError: expect.any(Function),
            }
          )
        })
      })

      describe('with Futures', () => {
        let cloned

        beforeEach(() => {
          cloned = React.cloneElement(test, {eventual: after(1, 5)})
          render.update(cloned)
        })

        it('is called for a resolved with undefined onError', async () => {
          expect(reducer).not.toHaveBeenCalled()
          instance.trigger()
          expect(reducer).not.toHaveBeenCalled()
          await wait(2)
          expect(reducer).toHaveBeenCalledWith(
            5,
            {loading: false},
            {eventual: expect.any(future), reducer}
          )
        })

        it('is called for a resolved with defined onError', async () => {
          render.update(
            React.cloneElement(cloned, {
              onError: () => null,
            })
          )

          expect(reducer).not.toHaveBeenCalled()
          instance.trigger()
          expect(reducer).not.toHaveBeenCalled()
          await wait(2)
          expect(reducer).toHaveBeenCalledWith(
            5,
            {loading: false},
            {
              eventual: expect.any(future),
              reducer,
              onError: expect.any(Function),
            }
          )
        })
      })
    })

    describe('setState', () => {
      const reducer = jest.fn((v, prevState, props) => ({
        v,
        loading: !prevState.loading,
        props,
      }))

      describe('with Promises', () => {
        const test = (
          <Test eventual={() => Promise.resolve(5)} reducer={reducer} />
        )
        let instance
        let setStateSpy

        beforeEach(() => {
          reducer.mockClear()
          instance = renderer.create(test).getInstance()
          setStateSpy = jest.spyOn(instance, 'setState')
        })

        it('is called with correct partialState for a resolved', async () => {
          expect(setStateSpy).not.toHaveBeenCalled()
          instance.trigger()
          expect(setStateSpy).not.toHaveBeenCalled()
          await wait(2)
          expect(setStateSpy).toHaveBeenCalled()
          expect(instance.state).toEqual({
            v: 5,
            loading: true,
            props: expect.any(Object),
          })
        })
      })

      describe('with Futures', () => {
        const test = <Test eventual={after(1, 5)} reducer={reducer} />
        let instance
        let setStateSpy

        beforeEach(() => {
          reducer.mockClear()
          instance = renderer.create(test).getInstance()
          setStateSpy = jest.spyOn(instance, 'setState')
        })

        it('is called with correct partialState for a resolved', async () => {
          expect(setStateSpy).not.toHaveBeenCalled()
          instance.trigger()
          expect(setStateSpy).not.toHaveBeenCalled()
          await wait(2)
          expect(setStateSpy).toHaveBeenCalled()
          expect(instance.state).toEqual({
            v: 5,
            loading: true,
            props: expect.any(Object),
          })
        })
      })
    })

    describe('cancelling', () => {
      const p = jest.fn(async () => {
        await wait(10)
        return 5
      })
      const consoleSpy = jest.spyOn(console, 'error')
      const test = <Test eventual={p} reducer={() => ({loading: true})} />
      let count = 0

      let render
      let instance
      let setStateSpy

      beforeEach(() => {
        p.mockClear()
        consoleSpy.mockClear()
        render = renderer.create(test)
        instance = render.getInstance()
        instance.constructor.displayName = `TestWithResetWarnings-${(count += 1)}`
        setStateSpy = jest.spyOn(instance, 'setState')
      })

      describe('raw setState', () => {
        it('triggers the React warning', () => {
          render.unmount()
          expect(consoleSpy).not.toHaveBeenCalled()
          instance.setState()
          expect(consoleSpy).toHaveBeenCalled()
          expect(consoleSpy.mock.calls[0][0]).toMatchSnapshot()
        })

        it('triggers the React warning with async', async () => {
          p().then(v => instance.setState({v}))
          await wait(2)
          render.unmount()
          expect(setStateSpy).not.toHaveBeenCalled()
          expect(consoleSpy).not.toHaveBeenCalled()
          await wait(10)
          expect(setStateSpy).toHaveBeenCalledWith({v: 5})
          expect(consoleSpy).toHaveBeenCalled()
          expect(consoleSpy.mock.calls[0][0]).toMatchSnapshot()
        })
      })

      describe('setFutureState', () => {
        it('doesnâ€™t trigger the React warning', async () => {
          instance.trigger()
          await wait(2)
          render.unmount()
          expect(setStateSpy).not.toHaveBeenCalled()
          expect(consoleSpy).not.toHaveBeenCalled()
          await wait(10)
          expect(setStateSpy).not.toHaveBeenCalled()
          expect(consoleSpy).not.toHaveBeenCalled()
        })
      })
    })
  })
})
