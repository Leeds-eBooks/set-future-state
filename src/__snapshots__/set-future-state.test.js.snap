// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setFutureState cancelling raw setState triggers the React warning 1`] = `
"Warning: Can only update a mounted or mounting component. This usually means you called setState, replaceState, or forceUpdate on an unmounted component. This is a no-op.

Please check the code for the TestWithResetWarnings-1 component."
`;

exports[`setFutureState cancelling raw setState triggers the React warning with async 1`] = `
"Warning: Can only update a mounted or mounting component. This usually means you called setState, replaceState, or forceUpdate on an unmounted component. This is a no-op.

Please check the code for the TestWithResetWarnings-2 component."
`;

exports[`setFutureState throws for an incorrect first argument 1`] = `
"The first argument to setFutureState() must be a Future or a Function which returns a Promise.

Please check the code for the WithFutureState(TestBase) component."
`;
